java/compaction-core/src/main/java/sleeper/compaction/job/status/CompactionJobStatusesBuilder.java

```diff
@@ -26,25 +27,34 @@
 public class CompactionJobStatusesBuilder {
 
     private final Map<String, CompactionJobCreatedStatus> createdById = new TreeMap<>(); // Order by job ID for output
-    private final Map<String, CompactionJobStartedStatus> startedById = new HashMap<>();
-    private final Map<String, CompactionJobFinishedStatus> finishedById = new HashMap<>();
+    private final Map<String, List<CompactionJobRun>> jobRunsById = new HashMap<>();
     private final Map<String, Instant> expiryDateById = new HashMap<>();
+    private final Map<String, Map<String, CompactionJobStartedStatus>> startedStatusMap = new HashMap<>();
 
     public CompactionJobStatusesBuilder jobCreated(
             String jobId, CompactionJobCreatedStatus statusUpdate) {
         createdById.put(jobId, statusUpdate);
+        jobRunsById.put(jobId, new ArrayList<>());
         return this;
     }
 
     public CompactionJobStatusesBuilder jobStarted(
-            String jobId, CompactionJobStartedStatus statusUpdate) {
-        startedById.put(jobId, statusUpdate);
+            String jobId, CompactionJobStartedStatus statusUpdate, String taskId) {
+        if (!startedStatusMap.containsKey(jobId)) {
+            startedStatusMap.put(jobId, new HashMap<>());
+        }
+        if (startedStatusMap.get(jobId).containsKey(taskId)) {
+            jobRunsById.get(jobId).add(CompactionJobRun.started(taskId, startedStatusMap.get(jobId).remove(taskId)));
+        }
+        startedStatusMap.get(jobId).put(taskId, statusUpdate);
         return this;
     }
 
     public CompactionJobStatusesBuilder jobFinished(
-            String jobId, CompactionJobFinishedStatus statusUpdate) {
-        finishedById.put(jobId, statusUpdate);
+            String jobId, CompactionJobFinishedStatus statusUpdate, String taskId) {
+        if (startedStatusMap.containsKey(jobId) && startedStatusMap.get(jobId).containsKey(taskId)) {
+            jobRunsById.get(jobId).add(CompactionJobRun.finished(taskId, startedStatusMap.get(jobId).remove(taskId), statusUpdate));
+        }
         return this;
     }
```

Comment:
What will happen if the run status updates come back from the database out of order? Can that be tested please?